### PR TITLE
Adding resources param to ds py log model

### PR DESCRIPTION
**Pull request by lmoros-DB**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/lmoros-DB/mlflow/pull/13362?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/13362/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 13362
```

</p>
</details>

### What changes are proposed in this pull request?

Adding the "resources" param to the DSPy log_model and save_model functions.  This allows deploying endpoints without having to specified environmental variables: DATABRICKS_TOKEN, DATABRICKS_HOST, DATABRICKS_API_KEY, and DATABRICKS_API_BASE 

### How is this PR tested?

- Manual tests:  registering and deploying a DSPy model without defining the environmental variables mentioned

Registering and deploying:
image
image

Testing endpoint and Review App:
image
image

### Does this PR require documentation update?

- [X ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ X] Yes. Give a description of this change to be included in the release notes for MLflow users.

This change is part of adding a new DSPy flavor.  It allow to define the Databricks resources (endpoints and Vector Search indexes) used by the DSPy program when it is registered as a model.  When the registered model is deployed in Databricks, no environmental variables need to be defined (ie. no need to define:  DATABRICKS_TOKEN, DATABRICKS_HOST, DATABRICKS_API_KEY, and DATABRICKS_API_BASE)

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [X] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [X] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [X] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`No` should be selected for new features and larger changes.

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [X] No (this PR will be included in the next minor release)

### CODE REVIEW

**Comment by lmoros-DB** [comment]
Cc @dbczumar, @chenmoneygithub

**Review by dbczumar — APPROVED** [review]
This is spectacular! Thanks @lmoros-DB !

**Comment by dbczumar** [comment]
@mlflow-automation autoformat

File: mlflow/dspy/save.py [hunk 1]
```diff
@@ -43,6 +43,7 @@
     _validate_and_prepare_target_save_path,
 )
 from mlflow.utils.requirements_utils import _get_pinned_requirement
+from mlflow.models.resources import Resource, _ResourceBuilder
```
Comment (harupy) [inline mlflow/dspy/save.py:46]: Adding this here should fix the error:

```
from pathlib import Path
```

File: mlflow/dspy/save.py [hunk 2]
```diff
@@ -84,6 +85,7 @@ def save_model(
     pip_requirements: Optional[Union[List[str], str]] = None,
     extra_pip_requirements: Optional[Union[List[str], str]] = None,
     metadata: Optional[Dict[str, Any]] = None,
+    resources: Optional[Union[str, List[Resource]]] = None,
```
Comment (harupy) [inline mlflow/dspy/save.py:88]:
```
    resources: Optional[Union[str, Path, List[Resource]]] = None,
```

if `Path` is supported

**Pull request closed without merging**